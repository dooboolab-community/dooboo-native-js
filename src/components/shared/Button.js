// @flow
import React, { Component } from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
} from 'react-native';
import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { ratio, colors } from '../../utils/Styles';

const styles = StyleSheet.create({
  btn: {
    backgroundColor: 'transparent',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: 320,
    height: 52,
    borderColor: 'white',

    alignItems: 'center',
    justifyContent: 'center',
  },
  btnDisabled: {
    backgroundColor: 'rgb(243,243,243)',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: 320,
    height: 52,
    borderColor: '#333',

    alignItems: 'center',
    justifyContent: 'center',
  },
  txt: {
    fontSize: 14,
    color: 'white',
  },
  imgLeft: {
    width: 24,
    height: 24,
    position: 'absolute',
    left: 16,
  },
});

type Props = {
  isLoading?: boolean;
  isDisabled?: boolean;
  onPress?: () => void;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  imgLeftSrc?: any;
  imgLeftStyle?: ImageStyle;
  indicatorColor?: string;
  activeOpacity?: number;
  children?: any;
}

type State = {

}

class Button extends Component<Props, State> {
  static defaultProps: Props = {
    isLoading: false,
    isDisabled: false,
    style: styles.btn,
    textStyle: styles.txt,
    imgLeftStyle: styles.imgLeft,
    indicatorColor: 'white',
    activeOpacity: 0.5,
  };

  render() {
    if (this.props.isDisabled) {
      return (
        <View style={this.props.disabledStyle}>
          <Text style={this.props.textStyle}>{this.props.children}</Text>
        </View>
      );
    }
    if (this.props.isLoading) {
      return (
        <View style={this.props.style}>
          <ActivityIndicator size='small' color={this.props.indicatorColor} />
        </View>
      );
    }
    return (
      <TouchableOpacity
        activeOpacity={this.props.activeOpacity}
        onPress={this.props.onPress}
      >
        <View style={this.props.style}>
          {
            this.props.imgLeftSrc
              ? <Image
                style={this.props.imgLeftStyle}
                source={this.props.imgLeftSrc}
              />
              : null
          }
          <Text style={this.props.textStyle}>{this.props.children}</Text>
        </View>
      </TouchableOpacity>
    );
  }
}

export default Button;
