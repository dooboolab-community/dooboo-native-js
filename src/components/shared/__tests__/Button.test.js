import 'react-native';
import * as React from 'react';
import Button from '../Button';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';

describe('Button', () => {
  it('renders without crashing', () => {
    const rendered = renderer.create(<Button />).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('component test', () => {
    const wrapper = shallow(
      <Button />,
    );

    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
      wrapper.setProps({ filled: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('simulate onPress', () => {
      let cnt = 1;
      const onPress = () => {
        cnt++;
      };

      wrapper.setProps({ onPress: () => onPress()});
      expect(wrapper).toMatchSnapshot();

      wrapper.first().props().onPress();
      expect(cnt).toBe(2);
    });
  });
});
