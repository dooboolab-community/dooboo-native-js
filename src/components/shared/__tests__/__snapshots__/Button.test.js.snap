// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test renders as expected 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "white",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 14,
        }
      }
    />
  </Component>
</TouchableOpacity>
`;

exports[`Button component test renders as expected 2`] = `
<TouchableOpacity
  activeOpacity={0.5}
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "white",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 14,
        }
      }
    />
  </Component>
</TouchableOpacity>
`;

exports[`Button component test simulate onPress 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  onPress={[Function]}
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "white",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 14,
        }
      }
    />
  </Component>
</TouchableOpacity>
`;

exports[`Button renders without crashing 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "white",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 52,
        "justifyContent": "center",
        "width": 320,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 14,
        }
      }
    />
  </View>
</View>
`;
